Rework command alias handling, step 1

dktapps opened the pull request:
This PR started out as an effort to decouple Command and CommandMap, but it's turned into a bit more than that.

A summary of changes:

## UX
- Added `cmdalias create`, `cmdalias delete` and `cmdalias list` commands
- `/help` now shows prefixed names such as `pocketmine:help`
- Prefixed command name (e.g. `pocketmine:help`) are now visible to Minecraft clients
- Permission denied messages are now able to show more useful context when e.g. checking subcommand permissions
- Multiple commands claiming an alias make the alias unusable (an error will be shown when used telling the user to pick from the namespaced names), instead of whichever plugin loaded last getting lucky

## API
- Added `CommandAliasMap`, which handles mapping of aliases to namespaced command IDs
- Added `CommandSender->getCommandAliasMap()` for user-specific aliases
- Added `CommandMap->getAliasMap()` for global fallback aliases
- `Command` no longer tracks its own registered aliases (now the job of `CommandMap`), breaking circular dependency
- Aliases must now be provided to `CommandMap->register()`
- Aliases can now be individually registered and unregistered without re-registering/unregistering the whole command using `CommandAliasMap` APIs
- Aliases are no longer namespaced, only the main command name (e.g. `pocketmine:?` is now gone while `pocketmine:help` still exists)
- `Command` now requires a `$namespace` parameter, which replaces the old `$fallbackPrefix` parameter of `register()`. It should be set to the name of the plugin.

Relevant issues
- #6508 
- #3371 - this PR doesn't implement storage, but allows configuration of per-user aliases during server runtime

dktapps committed:
First step in restructuring command alias handling
This decouples Command from CommandMap, and moves the burden of tracking registered
aliases to CommandMap. This allows lots of simplification, and also solves a few
weird usage message issues.

- Active (registered) aliases are now tracked via CommandMapEntry
- Commands overriding other commands' aliases now update the original command's registered alias list properly
- Command alias lists now include prefixed aliases
- Prefixed aliases are now included in command data provided to the client
- Server-side /pocketmine:help is now visible on the client
- Permission testing can now provide context that's more relevant to the command invocation - e.g. if a user doesn't have /time set permission, it'll now show a more specific message where previously it would just show that the permission for /time was denied
- User-specified label is now used for permission messages instead of command name - this is more consistent with user expectations
- /help can now see prefixed aliases
- Removed magic alias registration behaviour for VanillaCommand that I don't think anyone expected
- Aliases are now provided to CommandMap via register() parameters, instead of being retrieved from the provided Command
- Command->get/setAliases(), get/setLabel() and getName() are removed
- Command->getName() pushed down to PluginCommand, as it's only useful for CommandExecutors as a command ID and shouldn't be used anywhere else

dktapps committed:
Don't allow registering the same command instance twice

dktapps committed:
Add ability to register & unregister specific command aliases for existing commands

dktapps committed:
Merge branch 'major-next' into command-alias-handling

dktapps committed:
Merge branch 'major-next' into command-alias-handling

dktapps committed:
CS

NTT1906 commented on src/command/CommandMapEntry.php:
```diff
@@ -0,0 +1,48 @@
+
+	public function getUsage() : Translatable|string{
+		//TODO: usage messages ought to use user-specified alias, not command preferred
+		//command-preferred is confusing if the user used a different alias
```
maybe adding an alias input parameter to the getUsage() method.
also update the default command usage in the language file from:
> commands.tp.usage=/tp [target player] <destination player> OR /tp [target player] <x> <y> <z> [<y-rot> <x-rot>]

to:
> commands.tp.usage=/{%0} [target player] <destination player> OR /{%0} [target player] <x> <y> <z> [<y-rot> <x-rot>]

then, translate `{%0}` to the used alias?

dktapps replied:
Yes, it's just a lot of work

dktapps replied:
Clarification on this for posterity: It's in the plans to have usage messages generated from command signatures (once we have a proper arg parsing system), so it doesn't seem to make much sense to make a lot of mess with this crappy legacy system

NTT1906 commented:
![image](https://github.com/user-attachments/assets/136b4cf3-7217-495f-8d4d-2c420074ed54)
> `Prefixed aliases are now visible to Minecraft clients`

I don't like the look of it. Sending fully prefixed/namespaced commands to clients feels spammy. It should be cleaner by default.

1. The last registered command should take the name (without the namespace), except for help.
2. Only show namespace versions if more than 1 command with the same name.
3. Maybe add an option for the command to choose whether they wanted to show or not the namespace.
4. While the `pocketmine` namespace is consistent, using `pmmp` is a shorter, cleaner alternative.
5. Maybe a bad idea, as it might cause confusion, but by allowing namespace commands to still work silently—no hints, they can still be executed.

6. Yet another bad idea, and idk how it could work with the current command system in PMMP. Maybe introduce some kind of bundled command? Create a super command class where commands with the same name get bundled into a single command data. This allows for registering multiple overloads at the cost of losing command descriptions and needs a way to separate the command args when passing down the command args.

dktapps commented:
Yeah, I'm not sure about the fully-qualified names either. I think the way that those are handled probably needs to be rethought.
For example I don't like prefixing the aliases. Probably we should only prefix the primary label.

Longer term I think I want to have those move towards being internalised command IDs which are used by aliases to locate a command (see #3371), rather than being able to be directly invoked. But that's too big of a change for one PR.

dktapps committed:
Merge branch 'major-next' into command-alias-handling

dktapps commented:
> I don't like the look of it. Sending fully prefixed/namespaced commands to clients feels spammy.

To be honest this was more of a side effect of changing the tracking of the registered aliases. Previously the command registration wasn't aware of its own prefixed aliases, which lead to all kinds of interesting surprises for people trying to block commands by name rather than using permissions, since the aliases weren't visible in `/help` nor in client-side hints. (Not that they should have been doing that, but it's still surprising to have these hidden things.)

Filtering out the prefixed aliases is a bit problematic since we currently don't restrict aliases or fallback prefixes from containing `:`, so there's no foolproof way to tell them apart.

dktapps committed:
Merge branch 'major-next' into command-alias-handling

dktapps committed:
Merge branch 'major-next' into command-alias-handling

dktapps committed:
Don't register prefixed non-primary aliases
just the primary alias (as in, the command ID) is enough

dktapps committed:
Revert changes to command name

dktapps committed:
Make alias conflicts more sane

Now, neither command will win in an alias conflict, as both commands will become unusable by that alias.
This ensures that there's no alias flipflopping as we would've seen in the past, e.g. if two plugins
used the same alias but were loaded in a random order. This makes the behaviour safer for practical usage
since replaying a command will have more predictable results.

dktapps committed:
Fix PHPStan

dktapps committed:
Fixed PluginCommand usage messages

dktapps committed:
First look at user-local alias maps and /cmdalias

dktapps committed:
Suppress error
NetworkSession is a rat's nest of these sorts of errors. It'll need to be rewritten at some point.

dktapps committed:
Merge branch 'major-next' into command-alias-handling

dktapps committed:
Put namespace inside Command
this obviates the need for CommandMapEntry.

dktapps committed:
Tidy namespace & name handling & validation

probably these should be restricted further, but this is enough for now, since the original code didn't check these either

dktapps committed:
Merge branch 'major-next' into command-alias-handling

dktapps committed:
UX improvements & localisations

dktapps committed:
stfu
